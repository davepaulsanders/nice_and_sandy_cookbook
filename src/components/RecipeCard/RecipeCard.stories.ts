import type { Meta, StoryObj } from '@storybook/react';
 
import RecipeCard from './RecipeCard';
 
const meta: Meta<typeof RecipeCard> = {
  component: RecipeCard,
};
 
export default meta;
type Story = StoryObj<typeof RecipeCard>;
export const RecipeCards: Story = {
	args: {
	img: "https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg",
	href: "https://www.feastingathome.com/vegan-fajitas/",
	label: "Vegan Fajitas",
	alt: "test",
	}
};
