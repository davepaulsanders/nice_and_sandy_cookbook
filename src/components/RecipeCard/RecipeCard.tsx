import CardHeader from "../CardHeader/CardHeader"
import CardContent from "../CardContent/CardContent"
import tack from "/thumb-tack.webp" 
import cancelPin from "/thumb-tack-cancel.webp" 

const RecipeCard = (props) => {
	const pinRecipeToggle = () => {
		if (props.isPinned) {
			props.setPinnedRecipes(prev => prev.filter(recipe => recipe.props.id !== props.id))
		} else {
			props.setPinnedRecipes(prev => [<RecipeCard isPinned={true} setPinnedRecipes={props.setPinnedRecipes} img={props.img} href={props.href} label={props.label} alt={props.alt}/>, ...prev])
	
		}
	}
	return (
		<div className="relative bg-lightg flex flex-col justify-content-center \
		items-center shadow-lg rounded-md">
			<CardHeader {...props} />
			<CardContent {...props} />
			<img onClick={pinRecipeToggle} width="30px" height="30px" className="absolute top-5 right-5 hover:scale-125 hover:cursor-pointer" src={props.isPinned ? cancelPin : tack}/>
		</div>
	)
	}
export default RecipeCard
