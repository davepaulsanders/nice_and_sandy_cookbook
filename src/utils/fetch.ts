export const fetchData = async (url: string) => {
	const resp = await fetch(url)
	if (resp.status !== 200) throw new Error(`Request failed with ${resp.status}`)
	const data = resp.json()
	return data
}
