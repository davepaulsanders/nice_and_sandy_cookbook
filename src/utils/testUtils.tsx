import RecipeCard from '../components/RecipeCard/RecipeCard';
import { ReactElement } from "react"
import { vi } from "vitest"
import { Recipe } from "../types/types"

const generateRecipes = () => {
const recipes: Recipe[] = []
for (let i = 0; i < 10; i++) {
	recipes.push({
	id:i + 1,
	img:"https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg",
	href:"https://www.feastingathome.com/vegan-fajitas/",
	label:`Vegan Fajitas ${i + 1}`,
	alt:"test",
	category:"ENTREE",
	category_id:1,
	is_pinned: false,
	})
}
return recipes
}
const generateRecipeCards = () => {
const recipeCards: ReactElement[] = []
for (let i = 0; i < 10; i++) {
	recipeCards.push(
	<RecipeCard
	id={i + 1}
	img="https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg"
	href="https://www.feastingathome.com/vegan-fajitas/"
	label="Vegan Fajitas 2"
	alt="test"
	category="ENTREE"
	category_id={1}
	is_pinned= {false}
	recipes={generateRecipes()}
	setRecipes={vi.fn()}
	/>)
	}
	return recipeCards
}
export { generateRecipeCards, generateRecipes }
