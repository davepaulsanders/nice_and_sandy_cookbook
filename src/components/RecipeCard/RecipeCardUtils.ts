import { fetchData } from "../../utils/fetch"

export const pinRecipeToggle = async (e, props) => {
		e.preventDefault()
		const {recipes, setRecipes, is_pinned, ...rest} = props
		await fetchData(`../v1/recipes/pinned/${props.id}`, 
					{
						method: "PATCH",
						"Content-Type": "application/json"
				})
		if (props.is_pinned) {
			props.setRecipes(prev => {
				return prev.map((rec) => {
				if (rec.id === props.id) {
				  return { ...rest, is_pinned: false};
				}
				return rec;
				})
			  })

		} else {
			props.setRecipes(prev => {
				return prev.map((rec) => {
				if (rec.id === props.id) {
				  return { ...rest, is_pinned: true};
				}
				return rec;
				})
			  })
		}
	}
