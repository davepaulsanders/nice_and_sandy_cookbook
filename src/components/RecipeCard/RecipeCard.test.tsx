import RecipeCard from "./RecipeCard";
import { expect, test, vi } from "vitest";
import { render } from "@testing-library/react";
import { generateRecipes } from "../../utils/testUtils";

import "@testing-library/dom";

test("<RecipeCard /> as link renders", () => {
	const { container } = render(
		<RecipeCard
			id={1}
			img="https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg"
			href="https://www.feastingathome.com/vegan-fajitas/"
			label="Vegan Fajitas"
			alt="test"
			category="ENTREE"
			category_id={1}
			is_pinned={false}
			recipes={generateRecipes()}
			setRecipes={vi.fn()}
		/>
	);
	expect(container).toMatchSnapshot();
});
test("<RecipeCard /> as submit button renders", () => {
	const { container } = render(
		<RecipeCard
			id={1}
			img="https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg"
			href="https://www.feastingathome.com/vegan-fajitas/"
			label="Vegan Fajitas"
			alt="test"
			category="ENTREE"
			category_id={1}
			is_pinned={false}
			recipes={generateRecipes()}
			setRecipes={vi.fn()}
		/>
	);
	expect(container).toMatchSnapshot();
});
