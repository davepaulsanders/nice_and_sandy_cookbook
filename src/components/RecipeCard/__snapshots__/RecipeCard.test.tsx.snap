// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RecipeCard /> as link renders 1`] = `
<div>
  <div
    class="relative bg-lightg flex flex-col justify-content-center \\   items-center shadow-lg rounded-xl"
  >
    <div
      class="mt-4 w-11/12"
    >
      <img
        alt="test"
        class="h-[27rem] w-full object-cover"
        src="https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg"
      />
    </div>
    <div
      class="m-4"
    >
      <a
        class="text-xl hover:underline hover:cursor-pointer whitespace-nowrap"
        href="https://www.feastingathome.com/vegan-fajitas/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Vegan Fajitas
      </a>
    </div>
    <img
      class="absolute top-5 right-5 hover:scale-125 hover:cursor-pointer"
      height="30px"
      src="/thumb-tack.webp"
      width="30px"
    />
  </div>
</div>
`;

exports[`<RecipeCard /> as submit button renders 1`] = `
<div>
  <div
    class="relative bg-lightg flex flex-col justify-content-center \\   items-center shadow-lg rounded-xl"
  >
    <div
      class="mt-4 w-11/12"
    >
      <img
        alt="test"
        class="h-[27rem] w-full object-cover"
        src="https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg"
      />
    </div>
    <div
      class="m-4"
    >
      <a
        class="text-xl hover:underline hover:cursor-pointer whitespace-nowrap"
        href="https://www.feastingathome.com/vegan-fajitas/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Vegan Fajitas
      </a>
    </div>
    <img
      class="absolute top-5 right-5 hover:scale-125 hover:cursor-pointer"
      height="30px"
      src="/thumb-tack.webp"
      width="30px"
    />
  </div>
</div>
`;
